### PR TITLE
Add zeroed ends to parametric pulses

### PR DESCRIPTION
**Pull request by zachschoenfeld33**
## Summary

Zeros the ends of parametric pulses (rescaling the amplitude accordingly in the process). This avoids steep jumps in pulse edges.

NOTE: The zeroing occurs at `t=-1, t=duration+1`.

### Details and comments
- Tests updated, including testing the entire gaussian square array

### CODE REVIEW

File: test/python/pulse/test_pulse_lib.py [hunk 1]
```diff
@@ -127,16 +145,21 @@ def test_gauss_square_samples(self):
...
         amp = 0.5j
+        width = 100
         # formulaic
-        times = np.array(range(25), dtype=np.complex_)
-        times = times - (25 / 2) + 0.5
-        gauss = amp * np.exp(-(times / sigma)**2 / 2)
+        gauss_duration = duration - width
+        times = np.array(range(gauss_duration), dtype=np.complex_)
+        times = times - (gauss_duration / 2) + 0.5
+        gauss, _ = self._zero_ends_gaussian(times, amp, sigma, gauss_duration)
```
Comment (taalexander) [inline test/python/pulse/test_pulse_lib.py:153]: Why do you need to call this private method? Is there a way around this?

Comment (zachschoenfeld33) [reply test/python/pulse/test_pulse_lib.py:153]: Well I could call the method from `continuous.py`, but it seemed the idea of these tests was to avoid using anything from `pulse_lib` (since that's exactly how samples are generated).

I could write this logic in each test, but I need it in multiple tests so I thought using the function would be better practice.

File: qiskit/pulse/pulse_lib/parametric_pulses.py [hunk 2]
```diff
@@ -26,7 +26,8 @@
   - be a well known and/or well described formula (include the formula in the class docstring)
   - take some parameters (at least `duration`) and validate them, if necessary
   - implement a `get_sample_pulse` method which returns a corresponding SamplePulse in the
-    case that it is assembled for a backend which does not support it.
+    case that it is assembled for a backend which does not support it. Ends are zeroed to more
+    closely reflect the behavior of the backend.
```
Comment (nkanazawa1989) [inline qiskit/pulse/pulse_lib/parametric_pulses.py:30]: I think qiskit pulse is backend agnostic. Can we just say like zeroed to avoid steep jump in pulse edges?

**Comment by nkanazawa1989** [comment]
Hi Zach, thanks for the update. I think this change is really important. Could you please write reno as well?

close #4317

File: test/python/pulse/test_pulse_lib.py (hunk 1 above)
Comment (taalexander) [reply test/python/pulse/test_pulse_lib.py:153]: I think its fine to call this function, or just make sure the ends of the pulse are zero.

Comment (zachschoenfeld33) [reply test/python/pulse/test_pulse_lib.py]: Fixed in https://github.com/Qiskit/qiskit-terra/pull/4640/commits/e2082984e0793392b7a0d257fd8ed6b6d0a4a20f

File: qiskit/pulse/pulse_lib/parametric_pulses.py (hunk 2 above)
Comment (zachschoenfeld33) [reply qiskit/pulse/pulse_lib/parametric_pulses.py]: Fixed in https://github.com/Qiskit/qiskit-terra/pull/4640/commits/e2082984e0793392b7a0d257fd8ed6b6d0a4a20f.

File: qiskit/pulse/library/continuous.py [hunk 3]
```diff
@@ -150,20 +150,20 @@ def _fix_gaussian_width(gaussian_samples, amp: float, center: float, sigma: floa
...
         $\Omega_g(center \pm zeroed_width/2)=0$ is satisfied. This is used to avoid
         large discontinuities at the start of a gaussian pulse. If unsupplied,
-        defaults to $2*center$ such that $\Omega_g(0)=0$ and $\Omega_g(2*center)=0$.
+        defaults to $2*center + 1$ such that $\Omega_g(-1)=0$ and $\Omega_g(2*center + 1)=0$.
     rescale_amp: If True the pulse will be rescaled so that $\Omega_g(center)=amp$.
     ret_scale_factor: Return amplitude scale factor.
     """
     if zeroed_width is None:
-        zeroed_width = 2*center
+        zeroed_width = 2*center + 1
```
Comment (taalexander) [inline qiskit/pulse/library/continuous.py:166]: Shouldn't this be? 
```suggestion
        zeroed_width = 2*(center + 1)
```

File: qiskit/pulse/library/continuous.py [hunk 4]
```diff
@@ -236,20 +236,20 @@ def _fix_sech_width(sech_samples, amp: float, center: float, sigma: float,
...
         $\Omega_g(center \pm zeroed_width/2)=0$ is satisfied. This is used to avoid
         large discontinuities at the start of a sech pulse. If unsupplied,
-        defaults to $2*center$ such that $\Omega_g(0)=0$ and $\Omega_g(2*center)=0$.
+        defaults to $2*center + 1$ such that $\Omega_g(-1)=0$ and $\Omega_g(2*center + 1)=0$.
     rescale_amp: If True the pulse will be rescaled so that $\Omega_g(center)=amp$.
     ret_scale_factor: Return amplitude scale factor.
     """
     if zeroed_width is None:
-        zeroed_width = 2*center
+        zeroed_width = 2*center + 1
```
Comment (taalexander) [inline qiskit/pulse/library/continuous.py:252]: Same here.

File: qiskit/pulse/library/continuous.py (hunk 3 above)
Comment (taalexander) [reply qiskit/pulse/library/continuous.py:166]: Perhaps the calculation of `zero_offset` below is incorrect?

Comment (zachschoenfeld33) [reply qiskit/pulse/library/continuous.py]: Addressed in https://github.com/Qiskit/qiskit-terra/pull/4640/commits/ac3fdc7aa1f73a2ea496bec88b8e6b03b28f1e97

File: qiskit/pulse/library/continuous.py (hunk 4 above)
Comment (zachschoenfeld33) [reply qiskit/pulse/library/continuous.py]: Addressed in https://github.com/Qiskit/qiskit-terra/pull/4640/commits/ac3fdc7aa1f73a2ea496bec88b8e6b03b28f1e97.

File: qiskit/pulse/library/continuous.py [hunk 5]
```diff
@@ -158,14 +158,15 @@ def _fix_gaussian_width(gaussian_samples, amp: float, center: float, sigma: floa
...
     rescale_amp: If True the pulse will be rescaled so that $\Omega_g(center)=amp$.
     ret_scale_factor: Return amplitude scale factor.
     """
     if zeroed_width is None:
-        zeroed_width = 2*center + 1
+        zeroed_width = 2*(center + 1)
 
-    zero_offset = gaussian(np.array([zeroed_width/2]), amp, 0, sigma)
+    # compute the offset using ``zeroed_width-1`` (``zeroed_width`` includes both endpoints)
+    zero_offset = gaussian(np.array([(zeroed_width-1)/2]), amp, 0, sigma)
```
Comment (taalexander) [inline qiskit/pulse/library/continuous.py:169]: Why was this changed?

Comment (zachschoenfeld33) [reply qiskit/pulse/library/continuous.py]: My mistake. Fixed in https://github.com/Qiskit/qiskit-terra/pull/4640/commits/8b0c388cfc654b0aa78a1001d770ea1480b5203c.

File: test/python/pulse/test_pulse_lib.py [hunk 6]
```diff
@@ -106,38 +106,71 @@ def test_sampled_pulse(self):
...
         """Test that the gaussian samples match the formula."""
         duration = 25
         sigma = 4
         amp = 0.5j
-        # formulaic
-        times = np.array(range(25), dtype=np.complex_)
-        times = times - (duration / 2) + 0.5
+
+        # formulaic (with fixed zeroed endpoints at ``t=-1, duration+1``)
+        times = -1.0 + np.array(range(duration+2), dtype=np.complex_)
```
Comment (taalexander) [inline test/python/pulse/test_pulse_lib.py:119]: These tests are just about impossible to understand.

File: test/python/pulse/test_pulse_lib.py [hunk 7]
```diff
@@ -106,38 +106,71 @@ def test_sampled_pulse(self):
...
+        times[duration+1] += 1  # fix duration+1 endpoint
+
         gauss = amp * np.exp(-(times / sigma)**2 / 2)
+        zero_offset = gauss[0]
+        gauss -= zero_offset
+        amp_scale_factor = amp/(amp-zero_offset) if amp-zero_offset != 0 else 1
+        gauss *= amp_scale_factor
+
+        # verify zero offset and rescaling
+        self.assertEqual(gauss[0], 0.0)
```
Comment (taalexander) [inline test/python/pulse/test_pulse_lib.py:131]: Why are you testing the samples you created in this test?, you should be testing `samples`. Then almost all of the above is unnecessary. Use `library.continuous.gaussian` to generate this samples should be fine. These comments apply to the tests below as well.

File: qiskit/pulse/library/continuous.py [hunk 8]
```diff
@@ -214,17 +214,24 @@ def gaussian(times: np.ndarray, amp: complex, center: float, sigma: float,
 
 
 def gaussian_deriv(times: np.ndarray, amp: complex, center: float, sigma: float,
+                   zeroed_width: Optional[float] = None, rescale_amp: bool = False,
```
[the remark this answers is not in the document]
Comment (zachschoenfeld33) [reply qiskit/pulse/library/continuous.py]: Addressed in https://github.com/Qiskit/qiskit-terra/pull/4640/commits/8f427d9c326edf952247105d5392d67581d8bc5d

File: qiskit/pulse/library/parametric_pulses.py [hunk 9]
```diff
@@ -25,8 +25,10 @@
   - have a descriptive name
   - be a well known and/or well described formula (include the formula in the class docstring)
   - take some parameters (at least `duration`) and validate them, if necessary
-  - implement a `get_sample_pulse` method which returns a corresponding Waveform in the
-    case that it is assembled for a backend which does not support it.
+  - implement a `get_sample_pulse` method which returns a corresponding SamplePulse in the
```
[the remark this answers is not in the document]
Comment (zachschoenfeld33) [reply qiskit/pulse/library/parametric_pulses.py]: Addressed in https://github.com/Qiskit/qiskit-terra/pull/4640/commits/8f427d9c326edf952247105d5392d67581d8bc5d.

File: test/python/pulse/test_pulse_lib.py (hunk 6 above)
Comment (zachschoenfeld33) [reply test/python/pulse/test_pulse_lib.py]: Addressed in https://github.com/Qiskit/qiskit-terra/pull/4640/commits/8f427d9c326edf952247105d5392d67581d8bc5d.

File: test/python/pulse/test_pulse_lib.py (hunk 7 above)
Comment (zachschoenfeld33) [reply test/python/pulse/test_pulse_lib.py]: Addressed in https://github.com/Qiskit/qiskit-terra/pull/4640/commits/8f427d9c326edf952247105d5392d67581d8bc5d.
